openerp/addons/base/res/res_partner.py

```diff
@@ -331,9 +331,25 @@ def fields_view_get(self, cr, user, view_id=None, view_type='form', context=None
         'use_parent_address': False,
         'image': False,
     }
+    
+    def _check_company(self, cr, uid, ids, context=None):
+        if context == None:
+            context = {}
+        for partner in self.browse(cr, uid, ids,  context=context):
+            if (partner.parent_id and 
+                partner.parent_id.company_id and 
+                partner.company_id and 
+                partner.company_id.id !=partner.parent_id.company_id.id):
+                return False
+            return True
 
     _constraints = [
-        (osv.osv._check_recursion, 'You cannot create recursive Partner hierarchies.', ['parent_id']),
+        (osv.osv._check_recursion, 
+            'You cannot create recursive Partner hierarchies.', 
+            ['parent_id']),
+        (_check_company, 'You cannot create or update a child customer\
```

Comment:
En lugar de "create or update" quedaría mejor simplemente "set" y debe ser genérico porque no solo aplica para los clientes, entonces el mensaje quedaría así

'You cannot set a child partner with different company that the parent'

Ahora la prengunta, no quedaría mejor implementar que cuando se seleccione un contacto con una empresa el campo ```compan_id``` quedara oculto y establecido automáticamente al mismo valor que la empresa?

Creo que de esa forma empezaríamos a mejorar la usabilidad de la ficha de partners que tiene mucho que trabajarse.